tidy3d/components/geometry/base.py

```diff
@@ -1795,6 +1951,307 @@ def _surface_area(self, bounds: Bound) -> float:
+        y : np.ndarray[float]
+            Array of point positions in y direction.
+        z : np.ndarray[float]
+            Array of point positions in z direction.
+
+        Returns
+        -------
+        np.ndarray[bool]
+            ``True`` for every point that is inside the geometry.
+        """
+        x = np.array(x)
+        y = np.array(y)
+        z = np.array(z)
+        xyz = np.dot(self.inverse, np.vstack((x.flat, y.flat, z.flat, np.ones(x.size))))
+        return self.geometry.inside(xyz[0], xyz[1], xyz[2]).reshape(x.shape)
+
+    def _volume(self, bounds: Bound) -> float:
+        """Returns object's volume within given bounds."""
+        # NOTE (Lucas): Bounds are overestimated.
+        vertices = np.dot(self.inverse, self._vertices_from_bounds(bounds))[:3]
+        inverse_bounds = (tuple(vertices.min(axis=1)), tuple(vertices.max(axis=1)))
+        return abs(np.linalg.det(self.transform)) * self.geometry.volume(inverse_bounds)
+
+    def _surface_area(self, bounds: Bound) -> float:
+        """Returns object's surface area within given bounds."""
+        log.warning("Surface area of transformed elements cannot be calculated.")
+        return 0.0
+
+    @staticmethod
+    def translation(x: float, y: float, z: float) -> MatrixReal4x4:
+        """Return a translation matrix.
+
+        Parameters
+        ----------
+        x : float
+            Translation along x.
+        y : float
+            Translation along y.
+        z : float
+            Translation along z.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        return np.array(
+            [
+                (1.0, 0.0, 0.0, x),
+                (0.0, 1.0, 0.0, y),
+                (0.0, 0.0, 1.0, z),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def scaling(x: float, y: float = None, z: float = None) -> MatrixReal4x4:
+        """Return a scaling matrix.
+
+        Parameters
+        ----------
+        x : float
+            Scaling factor along x.
+        y : float = None
+            Scaling factor along y. If ``None``, use the value for x.
+        z : float = None
+            Scaling factor along z. If ``None``, use the value for x.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        if y is None:
+            y = x
+
+        if z is None:
+            z = x
+
+        if np.isclose((x, y, z), 0.0).any():
+            raise Tidy3dError("Scaling factors cannot be zero in any dimensions.")
+        return np.array(
+            [
+                (x, 0.0, 0.0, 0.0),
+                (0.0, y, 0.0, 0.0),
+                (0.0, 0.0, z, 0.0),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def rotation(angle: float, axis: Axis) -> MatrixReal4x4:
```

Comment:
add a staticmethod for rotation around any axis given by a vector?

Reply:
I didn't add it because of our weird axis convention: if the user specifies the y-axis using the general rotation vs the current `axis=1` option, they end results will be rotation on opposite directions. Is that ok?

Reply:
we have this, just FYI, not sure if it helps.
Seems some of the rotation bits added to help support anisotropic medium could be useful in here? but not sure.

https://github.com/flexcompute/tidy3d/blob/f60bf47f84264bc622e548fa41dfc344fb85ceaf/tidy3d/components/geometry/base.py#L542-L555

Reply:
I did see that, but here I need a 4×4 matrix for homogeneous coordinates (and originally only rotations arouns the 3 main axes). We can merge both by adding, for example, another parameter to `RotationAroundAxis(..., homogeneous_coordinates: bool)`. Then, create 2 more classes for `Translation` and `Scaling`.
TBH I don't see much value in creating classes fro those, there's nothing special there, just matrices, but let me know if you prefer it.

Reply:
I've pushed a commit where the `rotation` method uses the `RotationAroundAxis` class internally. The question of how to handle homogeneous coordinates and to replace all transforms with classes remains.

Reply:
I'm less concerned about creating classes for these, but more just suggesting we have a consistent implementation of rotation if possible and therefore don't have to maintain several versions. I think using RotationAroundAxis internally is good for me on this, thanks!